Feature: connect Exante

alexander-myltsev opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

#### Description
<!--- Describe your changes in detail -->
Add Exante to LEAN

<!--- #### Related Issue -->
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Implement live trading using Exante brokerage account

<!---#### Requires Documentation Change -->
<!--- Please indicate if these changes will require updates to documentation, and if so, specify what changes are required -->

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Tests are at https://github.com/alexander-myltsev/Lean/tree/feature-exante_connect/Tests/Brokerages/Exante .
The code has been tested both in demo and live mode.

#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:
<!--- The following is a checklist of items that MUST be completed before a PR is accepted -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. <!--- If not applicable, please explain why -->
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

<!--- Template inspired by https://www.talater.com/open-source-templates/#/page/99 -->

jaredbroad commented on Brokerages/Exante/ExanteBrokerage.DataQueueHandler.cs:
```diff
@@ -0,0 +1,249 @@
+        /// </summary>
+        /// <param name="ets">Exante tick short stream data object</param>
+        /// <returns>LEAN Tick object</returns>
+        private Tick CreateTick(ExanteTickShort ets)
```
Please replace all contracted variable names with a full readable notation

jaredbroad commented on Brokerages/Exante/ExanteBrokerage.DataQueueHandler.cs:
```diff
@@ -0,0 +1,249 @@
+            }
+
+            // Tradier trades are US NY time only. Convert local server time to NY Time:
+            var utc = ets.Date;
```
Comment

jaredbroad commented on Brokerages/Exante/ExanteBrokerage.DataQueueHandler.cs:
```diff
@@ -0,0 +1,249 @@
+            Aggregator.Remove(dataConfig);
+        }
+
+        /// <inheritdoc />
```
Our code are docs by policy we fully implement all comments; please copy paste if required but fully implement each XML block.

jaredbroad commented on Brokerages/Exante/ExanteBrokerage.cs:
```diff
@@ -0,0 +1,390 @@
+                case DayTimeInForce _:
+                    orderDuration = ExanteOrderDuration.Day;
+                    break;
+                case GoodTilDateTimeInForce gtdtif:
+                    orderDuration = ExanteOrderDuration.GoodTillTime;
+                    gttExpiration = gtdtif.Expiry;
```
Please fully expand all variable names

jaredbroad commented on Brokerages/QuantConnect.Brokerages.csproj:
```diff
@@ -5,7 +5,7 @@
     <AssemblyName>QuantConnect.Brokerages</AssemblyName>
     <TargetFramework>net462</TargetFramework>
-    <LangVersion>6</LangVersion>
+    <LangVersion>7.3</LangVersion>
```
To prevent backwards compatibility and cloud conflicts please leave this at v6,

alexander-myltsev replied:
I use features as follows

![image](https://user-images.githubusercontent.com/574267/123557183-1f982e80-d798-11eb-89e1-f72b5c206f37.png)

The code will become longer without them. I'm not sure exactly what do you mean by compatibility (you could bump up dependant libraries) and cloud conflict (make no sense). Please, confirm you're 100% tightened to v6 and will roll it back.

jaredbroad replied:
Correct. Its required - please leave at v6.

Martin-Molinero replied:
Hey! Seems the PR needs a rebase, we currently target net5.0 and are using language v9 in all Lean

alexander-myltsev replied:
It requires rebasing indeed. Alas, the blocker is https://github.com/QuantConnect/Lean/pull/5692#issuecomment-869214780 . I'll make another attempt and let you know.

jaredbroad commented on Common/Brokerages/ExanteBrokerageModel.cs:
```diff
@@ -0,0 +1,78 @@
+        }
+
+        /// <summary>
+        /// Bitfinex global leverage rule
```
Comment?

jaredbroad commented on Common/Extensions.cs:
```diff
@@ -650,6 +650,24 @@ public static string LazyToUpper(this string data)
+        /// </summary>
+        /// <param name="data">The string to lower</param>
+        /// <returns>The lower string</returns>
+        public static string LazyToLower(this string data)
```
Interesting 👌

jaredbroad commented on ToolBox/ExanteDownloader/ExanteDataDownloader.cs:
```diff
@@ -0,0 +1,19 @@
+namespace QuantConnect.ToolBox.ExanteDownloader
+{
+    public class ExanteDataDownloader
+    {
+        /// <summary>
+        /// Creates a new instance of <see cref="ExanteDataDownloader"/>
+        /// </summary>
+        public ExanteDataDownloader()
+        {
+        }
+
+        /// <summary>
+        /// Downloads data from Exante data source
+        /// </summary>
+        public void Download()
+        {
+        }
+    }
+}
```
Please delete if not implementing

alexander-myltsev replied:
I implemented the stubs. Do you need any data from Exante? Alas, I don't know. I will delete them if not.

jaredbroad replied:
Ideal would be to implement them for the open-source community and so others using Exante locally could use them.

Martin-Molinero replied:
> Please delete if not implementing

Reminder

jaredbroad commented on ToolBox/ExanteDownloader/ExanteDownloaderProgram.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace QuantConnect.ToolBox.ExanteDownloader
+{
+    public class ExanteDownloaderProgram
```
Please delete if not implementing

jaredbroad commented on ToolBox/Program.cs:
```diff
@@ -195,6 +196,12 @@ public static void Main(string[] args)
                             toDate);
                         break;
+                    
+                    case "exntdl":
+                    case "exantedownloader":
+                        ExanteDownloaderProgram.ExanteDownloader(tickers, resolution, fromDate, toDate);
+                        break;
```
Please delete if not implementing

Martin-Molinero replied:
> Please delete if not implementing

Ping

jaredbroad reviewed (changes requested):
Overall a great start, there are a few style tweaks to comments and variables and then we'll need to confirm it passes tests but it's on its way 💯

alexander-myltsev commented:
@alexander-myltsev I asked you in Slack but got no answer. I dare to ask here :) 

I wonder where the price of the pending order comes from. For example, there is a code at `Brokerages/Binance/BinanceBrokerage.cs:186`:

```csharp
                    case "MARKET":
                        order = new MarketOrder { Price = item.Price };
                        break;
```

If the order is pending then where the price comes from? I suppose it’s undefined until the order is fulfilled.

alexander-myltsev commented:
I’m trying to rebase to the `master`. I’m getting lots of errors as follows:

```
  Failed ComposeDataQueueHandlerInstances [54 ms]
  Error Message:
   OneTimeSetUp: System.TypeInitializationException : The type initializer for 'Delegates' threw an exception.
  ----> System.MissingMethodException : Failed to load symbol PyDictProxy_New: dlsym(RTLD_DEFAULT, PyDictProxy_New): symbol not found
```

I guess I’m missing a dependency. I installed `pythonnet` but it didn’t help. Any hint on how to fix that?

alexander-myltsev commented:
@jaredbroad I also have a question about exchanges. For example, SPY is traded at ARCA (full ticker SPY.ARCA) and AAPL is at NASDAQ (full ticker AAPL.NASDAQ). I see only the `USA` market in QuantConnect source code. Should I hardcode the correspondence of a ticker and its exchange to SymbolMapper?

Martin-Molinero commented:
Hey!

> 
> 
> I’m trying to rebase to the `master`. I’m getting lots of errors as follows:
> 
> ```
>   Failed ComposeDataQueueHandlerInstances [54 ms]
>   Error Message:
>    OneTimeSetUp: System.TypeInitializationException : The type initializer for 'Delegates' threw an exception.
>   ----> System.MissingMethodException : Failed to load symbol PyDictProxy_New: dlsym(RTLD_DEFAULT, PyDictProxy_New): symbol not found
> ```
> 
> I guess I’m missing a dependency. I installed `pythonnet` but it didn’t help. Any hint on how to fix that?

You should not install `pythonnet`, doing so will cause issues, you will have to remove it.
Did you follow https://github.com/QuantConnect/Lean/tree/master/Algorithm.Python, sounds like missing to set `PYTHONNET_PYDLL`?

> 
> 
> @jaredbroad Jared,
> 
> I asked you in Slack but got no answer. I dare to ask here :)
> 
> I wonder where the price of the pending order comes from. For example, there is a code at `Brokerages/Binance/BinanceBrokerage.cs:186`:
> 
> ```cs
>                     case "MARKET":
>                         order = new MarketOrder { Price = item.Price };
>                         break;
> ```
> 
> If the order is pending then where the price comes from? I suppose it’s undefined until the order is fulfilled.

Market orders do not have any price until they fill. Limit orders for example do have a price which is provided by the user or by the brokerage implementation itself if the order existed prior to starting lean, see `GetOpenOrders()`

Martin-Molinero commented:
> 
> 
> @jaredbroad I also have a question about exchanges. For example, SPY is traded at ARCA (full ticker SPY.ARCA) and AAPL is at NASDAQ (full ticker AAPL.NASDAQ). I see only the `USA` market in QuantConnect source code. Should I hardcode the correspondence of a ticker and its exchange to SymbolMapper?

Does Exante require the exchange somehow? From what I see it's only fetched from the client when emitting Ticks 👌, not when placing orders. Maybe it could be cached once by the symbolMapper

alexander-myltsev commented:
@Martin-Molinero , 

sorry for the delay. 

> > @jaredbroad I also have a question about exchanges. For example, SPY is traded at ARCA (full ticker SPY.ARCA) and AAPL is at NASDAQ (full ticker AAPL.NASDAQ). I see only the `USA` market in QuantConnect source code. Should I hardcode the correspondence of a ticker and its exchange to SymbolMapper?
> 
> Does Exante require the exchange somehow? From what I see it's only fetched from the client when emitting Ticks 👌, not when placing orders. Maybe it could be cached once by the symbolMapper. Exante doesn't provide it with the data feed?

Let me clarify where my question goes from. To pass, for example, `ExanteBrokerageTests` I need to create a symbol as follows:

```c#
var securityType = SecurityType.Equity;
var market = "ARCA";
var ticker = "SPY";

Market.Add(market, 999);
SecurityType = securityType;
Symbol = Symbol.Create(ticker, securityType, market);
```

I can't simply use `QuantConnect.Tests.Symbols.SPY` since it has `Market.USA`. Then it goes to Exante symbol mapper, and I doubt how to resolve it. You propose to have a hash correspondence between a symbol name and an exchange. If there are two symbols with exact names at two different exchanges, then the idea doesn't work. Is there a exchange's requirement to have a unique name for the symbol?

Martin-Molinero commented:
> 
> 
> @Martin-Molinero ,
> 
> sorry for the delay.
> 
> > > @jaredbroad I also have a question about exchanges. For example, SPY is traded at ARCA (full ticker SPY.ARCA) and AAPL is at NASDAQ (full ticker AAPL.NASDAQ). I see only the `USA` market in QuantConnect source code. Should I hardcode the correspondence of a ticker and its exchange to SymbolMapper?
> > 
> > 
> > Does Exante require the exchange somehow? From what I see it's only fetched from the client when emitting Ticks 👌, not when placing orders. Maybe it could be cached once by the symbolMapper. Exante doesn't provide it with the data feed?
> 
> Let me clarify where my question goes from. To pass, for example, `ExanteBrokerageTests` I need to create a symbol as follows:
> 
> ```cs
> var securityType = SecurityType.Equity;
> var market = "ARCA";
> var ticker = "SPY";
> 
> Market.Add(market, 999);
> SecurityType = securityType;
> Symbol = Symbol.Create(ticker, securityType, market);
> ```
> 
> I can't simply use `QuantConnect.Tests.Symbols.SPY` since it has `Market.USA`. Then it goes to Exante symbol mapper, and I doubt how to resolve it. You propose to have a hash correspondence between a symbol name and an exchange. If there are two symbols with exact names at two different exchanges, then the idea doesn't work. Is there a exchange's requirement to have a unique name for the symbol?

For USA equities we've been successfully using Market.USA with no clash/issues, other markets will have a specific/different market. In which cases does Exante require the exchange which is causing issues for you?

> Does Exante require the exchange somehow? From what I see it's only fetched from the client when emitting Ticks 👌, not when placing orders. Maybe it could be cached once by the symbolMapper. Exante doesn't provide it with the data feed?

alexander-myltsev commented:
@Martin-Molinero please note that I rebased my commits to a more recent master.

---

> You should not install pythonnet, doing so will cause issues, you will have to remove it.
Did you follow https://github.com/QuantConnect/Lean/tree/master/Algorithm.Python, sounds like missing to set PYTHONNET_PYDLL?

That indeed was the reason. Please, note that Anaconda is kind of specific. Some people (including me) don't use it at all. I prefer `pyenv`. And to install python with "pydll" one should do it as follows: `PYTHON_CONFIGURE_OPTS="--enable-framework" pyenv install`. Somebody might find it useful.

--- 

I need to add more functionality related to symbols mapper. But I found that after rebase, launching even a single test takes minutes. I use IntelliJ Rider 2021.2. Do you know what the reason might be?

--- 

> For USA equities we've been successfully using Market.USA with no clash/issues, other markets will have a specific/different market. In which cases does Exante require the exchange which is causing issues for you?

OK. `Market.USA` seems working. But how to specify London or Frankfurt exchange to Lean?

---

Also, Exante keeps all cryptocurrencies under `EXANTE.<crypto symbol id>` identifier. They won't tell me what market they use  (if any) to deal with a particular cryptocurrency. How do you handle that in Lean?

Martin-Molinero commented on Brokerages/Exante/ExanteClientWrapper.cs:
```diff
@@ -0,0 +1,232 @@
+        public ExanteClientWrapper(ExanteClientOptions clientOptions)
+        {
+            _client = new ExanteClient(clientOptions);
+            StreamClient = new ExanteStreamClient(clientOptions);
```
Missing to implement `IDisposable` here and disposing of these, should be called by the brokerage implementation which should also implement `IDisposable` 👍

alexander-myltsev replied:
Done.

Martin-Molinero commented on Brokerages/Exante/ExanteBrokerage.cs:
```diff
@@ -0,0 +1,420 @@
+                        $"ExanteBrokerage.ConvertOrderType: Unsupported order type: {order.Type}");
+            }
+
+            foreach (var o in orderPlacement.Data)
```
There's a race condition between placing the order and getting the event through the callback `GetOrdersStreamAsync`, suggest using a `BrokerageConcurrentMessageHandler` instance to handle this, we've done this for multiple brokerages can see examples of it

alexander-myltsev replied:
Sorry, I don't see a race condition inside my code. Could you clarify? Does it happen inside `OnOrderEvent`?

Martin-Molinero replied:
> Sorry, I don't see a race condition inside my code. Please, explain where it comes from. Does it happen inside OnOrderEvent?

Sure, when you place a trade you add it to the `_orderMap` and the `GetOrdersStreamAsync` is reading from `_orderMap`. What if `GetOrdersStreamAsync` is triggered before the order is added to `_orderMap` it won't find it and not trigger the order event

alexander-myltsev replied:
> Sure, when you place a trade you add it to the _orderMap and the GetOrdersStreamAsync is reading from _orderMap. What if GetOrdersStreamAsync is triggered before the order is added to _orderMap it won't find it and not trigger the order event

Thanks for the explanation. You're right.

Martin-Molinero commented on Brokerages/Exante/ExanteBrokerage.cs:
```diff
@@ -0,0 +1,420 @@
+        private bool _isConnected;
+        private readonly ExanteClientWrapper _client;
+        private readonly string _accountId;
+        private readonly IDataAggregator Aggregator;
```
nit - `Aggregator` -> `_aggregator`

alexander-myltsev replied:
Done.

Martin-Molinero commented on Brokerages/Exante/ExanteBrokerage.cs:
```diff
@@ -0,0 +1,420 @@
+                Order order;
+                if (_orderMap.TryGetValue(exanteOrder.OrderId, out order))
+                {
+                    OnOrderEvent(new OrderEvent(order, DateTime.UtcNow, OrderFee.Zero) // TODO: What's the fee?
```
> // TODO: What's the fee?

If Exante does not provide it could directly use the FeeModel instance on the associated security, we've done this for a few brokerages too

alexander-myltsev replied:
I asked Exante some time ago. They haven't answered. I'll ask again if they provide.

Martin-Molinero commented on Brokerages/Exante/ExanteBrokerage.DataQueueHandler.cs:
```diff
@@ -0,0 +1,256 @@
+
+            var utc = exanteFeedTrade.Date;
+            // Convert the timestamp to exchange timezone and pass into algorithm
+            var time = utc.ConvertTo(DateTimeZone.Utc, TimeZones.NewYork);
```
We sure Exante is returning always UTC time? Should not convert to `TimeZones.NewYork` hard coded, should convert to the _exchange_ time zone, which changes by symbol, see for example `IB` `GetRealTimeTickTime(Symbol symbol)`

alexander-myltsev replied:
Seems you're right. I'll update.

Martin-Molinero replied:
ping

Martin-Molinero commented on Brokerages/Exante/ExanteBrokerage.DataQueueHandler.cs:
```diff
@@ -0,0 +1,256 @@
+            // Convert the timestamp to exchange timezone and pass into algorithm
+            var time = utc.ConvertTo(DateTimeZone.Utc, TimeZones.NewYork);
+
+            var instrument = _client.GetSymbol(symbolId);
```
Should probably cache this guy out `var instrument = _client.GetSymbol(symbolId);`? This might we called many times per second, doing this call will add a significant overhead

Martin-Molinero replied:
ping

Martin-Molinero commented on Brokerages/Exante/ExanteBrokerage.DataQueueHandler.cs:
```diff
@@ -0,0 +1,256 @@
+
+            var bids = exanteTickShort.Bid.ToList();
+            var asks = exanteTickShort.Ask.ToList();
+            return new Tick(exanteTickShort.Date, symbol, "", "",
```
Should follow the same patter for `exanteTickShort.Date` as for Trade data, similar comments here as above 👍

Martin-Molinero replied:
ping

Martin-Molinero commented on Common/Brokerages/ExanteBrokerageModel.cs:
```diff
@@ -0,0 +1,96 @@
+using System;
```
nit - missing license

Martin-Molinero replied:
ping

Martin-Molinero commented on Common/Brokerages/ExanteBrokerageModel.cs:
```diff
@@ -0,0 +1,96 @@
+        /// <returns>The benchmark for this brokerage</returns>
+        public override IBenchmark GetBenchmark(SecurityManager securities)
+        {
+            var symbol = Symbol.Create("SPY", SecurityType.Equity, Market.Bitfinex);
```
`"SPY", SecurityType.Equity, Market.Bitfinex` seems off?

Martin-Molinero replied:
ping

Martin-Molinero commented on Common/Orders/Fees/ExanteFeeModel.cs:
```diff
@@ -0,0 +1,112 @@
+
+                case SecurityType.Option:
+                case SecurityType.IndexOption:
+                    var optionsFee = ComputeOptionFee("USA", order);
```
nit - "USA" -> "Market.USA"

Martin-Molinero commented on Tests/Brokerages/Exante/ExanteBrokerageTests.cs:
```diff
@@ -0,0 +1,75 @@
+namespace QuantConnect.Tests.Brokerages.Exante
+{
+    [TestFixture]
+    public partial class ExanteBrokerageTests : BrokerageTests
```
This is still to be implemented right?

alexander-myltsev replied:
Some tests are derived from the base class. Some are added. All pass. I'll check from other classes what do you test. But if you comment on your requirements for tests, it would help.

alexander-myltsev replied:
@Martin-Molinero :arrow_up: ?

Martin-Molinero replied:
We've updated the https://github.com/QuantConnect/Lean.Brokerages.Template/tree/master/QuantConnect.TemplateBrokerage.Tests to include a nice set of expected tests (might depend on the brokerage implementation, supports IDQH etc)

Martin-Molinero reviewed (changes requested):
> @Martin-Molinero please note that I rebased my commits to a more recent master.

Ty, I still think you need to rebase a few commits more

> That indeed was the reason. Please, note that Anaconda is kind of specific. Some people (including me) don't use it at all. I prefer pyenv. And to install python with "pydll" one should do it as follows: PYTHON_CONFIGURE_OPTS="--enable-framework" pyenv install. Somebody might find it useful.

Great! That's for sharing 👍 

> I need to add more functionality related to symbols mapper. But I found that after rebase, launching even a single test takes minutes. I use JetBrains Rider 2021.2. Do you know what the reason might be?

Just pulled the branch and it worked correctly for me, using Rider 2021.1.5. Suggest `git clean -xqdf` will remove _anything_ left behind in the local repo. Might be a big in Rider 2021.2 too.
Might be missing to dispose/shutdown after each unit test?

> OK. Market.USA seems working. But how to specify London or Frankfurt exchange to Lean?

We will need to add a new Market 👍. But **I strongly suggest** doing it in steps. Getting the brokerage ready, shinning and shipped for USA Market is a great win for this PR.
A few standard test we manually run are

- Live Trading for 7 days with an algorithm
- Survive a internet outage


> Also, Exante keeps all cryptocurrencies under EXANTE.<crypto symbol id> identifier. They won't tell me what market they use (if any) to deal with a particular cryptocurrency. How do you handle that in Lean?

Not sure Exante allows trading cryptocurrencies directly? but they are underlying assets in some fund of them? https://exante.eu/press/news/1419-exante-jumps-on-the-crypto-rally-and-adds-more-coins-to-the-offering/ Don't see any crypto mention in their page https://exante.eu/markets/

alexander-myltsev commented on Tests/Brokerages/Exante/ExanteBrokersBrokerageDataQueueHandlerTest.cs:
```diff
@@ -0,0 +1,111 @@
+{
+    // Test DataQueueHandler
+    [TestFixture]
+    public partial class ExanteBrokerageTests
```
Done.

alexander-myltsev commented:
@Martin-Molinero Martin,

I rebased on the recent `master` branch.

---

> Not sure Exante allows trading cryptocurrencies directly? but they are underlying assets in some fund of them? https://exante.eu/press/news/1419-exante-jumps-on-the-crypto-rally-and-adds-more-coins-to-the-offering/ Don't see any crypto mention in their page https://exante.eu/markets/

They do. The list of supported crypto currencies: https://github.com/alexander-myltsev/Lean/blob/feature-exante_connect/Brokerages/Exante/ExanteBrokerage.cs#L63-L67.

--- 

> A few standard test we manually run are
> 
> Live Trading for 7 days with an algorithm
> Survive a internet outage

Which algorithm exactly? I'd like to run it on my side and check the source if it survives for 7 days.

Martin-Molinero commented:
Hey @alexander-myltsev! Gave the PR a new look and there seems to be quite a few reviews unaddressed/replied yet

Martin-Molinero commented:
Closing for now, reviews pending to be address.
 Moving forward new brokerage implementations should be done following a plugin approach. The core implementations should be done in a new repository using template https://github.com/QuantConnect/Lean.Brokerages.Template.
In Lean only there should only be the minimal required footprint to use a brokerage, for example https://github.com/QuantConnect/Lean/pull/5896 or https://github.com/QuantConnect/Lean/pull/5247

alexander-myltsev commented:
@Martin-Molinero Martin,

I checked again all review and fixed as follows:

> Does Exante require the exchange somehow? From what I see it's only fetched from the client when emitting Ticks 👌, not when placing orders. Maybe it could be cached once by the symbolMapper. Exante doesn't provide it with the data feed?

Fixed a long time ago.

---

> For USA equities we've been successfully using Market.USA with no clash/issues, other markets will have a specific/different market. In which cases does Exante require the exchange which is causing issues for you?

There are clashes. Some of the symbols are as follows: 

<details>
  <summary>Expand</summary>

```
20210902 18:47:39.185 ERROR:: Symbol SPY occurs on two exchanges: ARCA CBOE
20210902 18:47:39.185 ERROR:: Symbol NVDA occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.185 ERROR:: Symbol SPY occurs on two exchanges: ARCA CBOE
20210902 18:47:39.185 ERROR:: Symbol HBAN occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.185 ERROR:: Symbol AMD occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.185 ERROR:: Symbol TQQQ occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.185 ERROR:: Symbol EDIT occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.186 ERROR:: Symbol SRPT occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.186 ERROR:: Symbol LPSN occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.186 ERROR:: Symbol TSP occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.186 ERROR:: Symbol TSLA occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.186 ERROR:: Symbol NVDA occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.186 ERROR:: Symbol SNY occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.186 ERROR:: Symbol TGTX occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.186 ERROR:: Symbol BIIB occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.186 ERROR:: Symbol CRSR occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.186 ERROR:: Symbol XOP occurs on two exchanges: ARCA CBOE
20210902 18:47:39.186 ERROR:: Symbol OKTA occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.186 ERROR:: Symbol GOOGL occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.186 ERROR:: Symbol SPY occurs on two exchanges: ARCA CBOE
20210902 18:47:39.186 ERROR:: Symbol DIA occurs on two exchanges: ARCA CBOE
20210902 18:47:39.186 ERROR:: Symbol TGTX occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.186 ERROR:: Symbol TQQQ occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.186 ERROR:: Symbol INTC occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.186 ERROR:: Symbol OKTA occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.186 ERROR:: Symbol PAAS occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.186 ERROR:: Symbol NVDA occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.187 ERROR:: Symbol STNE occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.187 ERROR:: Symbol QQQ occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.187 ERROR:: Symbol NFLX occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.187 ERROR:: Symbol XLK occurs on two exchanges: ARCA CBOE
20210902 18:47:39.187 ERROR:: Symbol TECS occurs on two exchanges: ARCA CBOE
20210902 18:47:39.187 ERROR:: Symbol QQQ occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.187 ERROR:: Symbol ASML occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.187 ERROR:: Symbol SOXL occurs on two exchanges: ARCA CBOE
20210902 18:47:39.187 ERROR:: Symbol WISH occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.187 ERROR:: Symbol SPLG occurs on two exchanges: ARCA CBOE
20210902 18:47:39.278 ERROR:: Symbol SAVA occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.278 ERROR:: Symbol PRPL occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.278 ERROR:: Symbol KWEB occurs on two exchanges: ARCA CBOE
20210902 18:47:39.278 ERROR:: Symbol SDGR occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.278 ERROR:: Symbol HYG occurs on two exchanges: ARCA CBOE
20210902 18:47:39.278 ERROR:: Symbol BEAM occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.278 ERROR:: Symbol MDB occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.278 ERROR:: Symbol AVGO occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.278 ERROR:: Symbol IWB occurs on two exchanges: ARCA CBOE
20210902 18:47:39.278 ERROR:: Symbol GOOG occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.278 ERROR:: Symbol KARS occurs on two exchanges: ARCA CBOE
20210902 18:47:39.279 ERROR:: Symbol CHTR occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.279 ERROR:: Symbol OIH occurs on two exchanges: ARCA CBOE
20210902 18:47:39.279 ERROR:: Symbol TECL occurs on two exchanges: ARCA CBOE
20210902 18:47:39.279 ERROR:: Symbol ETSY occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.279 ERROR:: Symbol XLK occurs on two exchanges: ARCA CBOE
20210902 18:47:39.279 ERROR:: Symbol TNA occurs on two exchanges: ARCA CBOE
20210902 18:47:39.279 ERROR:: Symbol SGOC occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.279 ERROR:: Symbol BKNG occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.279 ERROR:: Symbol ADBE occurs on two exchanges: NASDAQ CBOE
20210902 18:47:39.279 ERROR:: Symbol BLNK occurs on two exchanges: NASDAQ CBOE
20210902 18:47:48.771 ERROR:: Symbol BLK occurs on two exchanges: CBOE CME
20210902 18:47:48.771 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.771 ERROR:: Symbol GE occurs on two exchanges: CBOE CME
20210902 18:47:48.771 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.771 ERROR:: Symbol GE occurs on two exchanges: CBOE CME
20210902 18:47:48.771 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.771 ERROR:: Symbol GE occurs on two exchanges: CBOE CME
20210902 18:47:48.771 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.771 ERROR:: Symbol GE occurs on two exchanges: CBOE CME
20210902 18:47:48.771 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.771 ERROR:: Symbol GE occurs on two exchanges: CBOE CME
20210902 18:47:48.772 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.772 ERROR:: Symbol GE occurs on two exchanges: CBOE CME
20210902 18:47:48.772 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.772 ERROR:: Symbol LE occurs on two exchanges: NASDAQ CME
20210902 18:47:48.772 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.772 ERROR:: Symbol GE occurs on two exchanges: CBOE CME
20210902 18:47:48.772 ERROR:: Symbol MBT occurs on two exchanges: CBOE CME
20210902 18:47:48.772 ERROR:: Symbol ETH occurs on two exchanges: USD CME
20210902 18:47:48.772 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.772 ERROR:: Symbol GE occurs on two exchanges: CBOE CME
20210902 18:47:48.772 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.772 ERROR:: Symbol GE occurs on two exchanges: CBOE CME
20210902 18:47:48.772 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.772 ERROR:: Symbol GE occurs on two exchanges: CBOE CME
20210902 18:47:48.772 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.773 ERROR:: Symbol GE occurs on two exchanges: CBOE CME
20210902 18:47:48.773 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.773 ERROR:: Symbol GE occurs on two exchanges: CBOE CME
20210902 18:47:48.773 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.773 ERROR:: Symbol GE occurs on two exchanges: CBOE CME
20210902 18:47:48.773 ERROR:: Symbol LE occurs on two exchanges: NASDAQ CME
20210902 18:47:48.773 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.773 ERROR:: Symbol GE occurs on two exchanges: CBOE CME
20210902 18:47:48.773 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.773 ERROR:: Symbol GE occurs on two exchanges: CBOE CME
20210902 18:47:48.773 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.773 ERROR:: Symbol GE occurs on two exchanges: CBOE CME
20210902 18:47:48.773 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.773 ERROR:: Symbol GE occurs on two exchanges: CBOE CME
20210902 18:47:48.774 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.774 ERROR:: Symbol GE occurs on two exchanges: CBOE CME
20210902 18:47:48.774 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.774 ERROR:: Symbol GE occurs on two exchanges: CBOE CME
20210902 18:47:48.774 ERROR:: Symbol BTC occurs on two exchanges: USD CME
20210902 18:47:48.774 ERROR:: Symbol GE occurs on two exchanges: CBOE CME
```

</details>

Errors would log at https://github.com/alexander-myltsev/Lean/blob/6bde3af5af45dbd4334b864c199ccf2853f692a3/Brokerages/Exante/ExanteBrokerage.cs#L141 

---

> Closing for now, reviews pending to be address.
> Moving forward new brokerage implementations should be done following a plugin approach. The core implementations should be done in a new repository using template https://github.com/QuantConnect/Lean.Brokerages.Template.
In Lean only there should only be the minimal required footprint to use a brokerage, for example #5896 or #5247

OK. I'm not sure I understand the implications. Does it mean that I need to move the code to another repository by implementing interfaces of a `Lean.Brokerages.Template`? And when I move it, the Exante plugin would be hosted by QuantConnect's Lean, right?

Martin-Molinero commented on ToolBox/Program.cs:
```diff
@@ -141,7 +141,7 @@ public static void Main(string[] args)
                             toDate);
                         break;
-
+                    
```
ping

Martin-Molinero reviewed (commented):
ping
